### PR TITLE
Added cjson.py which can be used to read cjson 

### PR DESCRIPTION
**Pull request by adityaomar3**
current method added is get_atoms_coords which will get the co-ordinates of all atoms this aims to solve https://github.com/OpenChemistry/avogadrolibs/issues/712

Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.

### CODE REVIEW

**Comment by adityaomar3** [comment]
currently it will work as 
`from avogadro import cjson as cj
cj.Cjson().get_atoms_coords("file_path")`

I will try to make it more simpler by day so that we dont need to create class object and will work simply like 
`from avogadro import cjson as cj
cj.get_atoms_coords("file_path")`

**Comment by adityaomar3** [comment]
@ghutchis Sir, can you please review it?

**Review by ghutchis — CHANGES_REQUESTED** [review]

**Comment by ghutchis** [comment]
Sorry, I'm not sure why my review comments were stuck as "pending" - I was wondering when you'd make the changes .. but you didn't get the comments in the first place.

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,6 @@
+{
+    "cSpell.words": [
```
[the remark this answers is not in the document]
Comment (adityaomar3) [reply .vscode/settings.json:2]: Yeah! Sure, I was also thinking to add these files to .gitignore.

File: python/avogadro/cjson.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    def __to_cjson(self, element_coords):
+        cjson_dict = {"element-coordinates" :element_coords}
+        cjsonData = json.dumps(cjson_dict, indent=4)
+        print(cjsonData)
+    def get_atoms_coords(self,filePath):
+        data = self.__open_file(filePath)
+        coords = data["atoms"]["coords"]["3d"]
+        elements = data["atoms"]["elements"]["number"]
+        element_coords = [(*coords[i*3:(i+1)*3], elements[i]) for i in range(0, int(len(coords) / 3))]
+        print(type(element_coords))
```
[the remark this answers is not in the document]
Comment (adityaomar3) [reply python/avogadro/cjson.py:18]: Oops! I forgot to remove that, will rectify this

**Comment by adityaomar3** [comment]
> not

Ah! Please don't be sorry, sometimes technical issues are there,I will rectify all mistakes shortly.

**Comment by adityaomar3** [comment]
@ghutchis I have done the changes as told. 
Whenever you get time, please review it again.

File: python/avogadro/cjson.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    def __open_file(self, filePath):
+        '''Use to read CJson formats by converting them to python dictionaries'''
+        with open(filePath, 'r') as cjsonFile:
+            py_dict_data = json.load(cjsonFile)
+            return py_dict_data
+    def __to_cjson(self, cjson_dict_file):
+        '''It converts python dictionaries to CJson format'''
+        cjsonData = json.dumps(cjson_dict_file, indent=4)
+        return (cjsonData)
+    def get_atoms_coords(self,filePath):
```
Comment (ghutchis) [inline python/avogadro/cjson.py:26]: I think we want both a method to get an array of coordinates and set the data["atoms"]["coords"]["3d"] from an array of (-1, 3) coordinates.

If you want to add this method (which returns coordinates + atomic number), I'd want to also see:
- get_atoms_coordinates() - this method
- set_atoms_coordinates() - setting the data[] from an array of coordinates and elements
- get_elements() - simply returning the elements from `data["atoms"]["elements"]["number"]`
- set_elements() - setting the elements
- get_coordinates() - just return the coordinates
- set_coordinates() - just set the coordinates

File: python/avogadro/cjson.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+            y co-ordinate
+            z co-ordinate
+            Atomic Number of Element
+        ]
+        """
+        data = self.__open_file(filePath)
+        coords = data["atoms"]["coords"]["3d"]
+        elements = data["atoms"]["elements"]["number"]
+        element_coords = [(*coords[i*3:(i+1)*3], elements[i]) for i in range(0, int(len(coords) / 3))]
+        cjson_dict = {"element-coordinates" :element_coords}
```
Comment (ghutchis) [inline python/avogadro/cjson.py:40]: Why? Shouldn't you just return the `element_coords` array?

File: python/avogadro/cjson.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+class Cjson:
+    """
+    This Class is intended to read Cjson files
+    with python libraries and perform certain
+    methods on files and convert them back to Cjson
+    files as required
+    """
+    def __init__(self):
+        pass
+    def __open_file(self, filePath):
```
Comment (ghutchis) [inline python/avogadro/cjson.py:17]: Generally when we're creating modules, we want symmetry in names. So if you're creating an `__open_file()` method, there should be a save version. So if you want `__to_cjson` then you want `__from_cjson`.

Or what if a user has a string (e.g. from a URL or from another source)?

So you might want to check if `filePath` can be considered a file path and if not, try loading it as a json string.

**Review by ghutchis — CHANGES_REQUESTED** [review]

File: python/avogadro/cjson.py (hunk 3 above)
Comment (adityaomar3) [reply python/avogadro/cjson.py:26]: Ok, I will add all these functions shortly.

File: python/avogadro/cjson.py (hunk 4 above)
Comment (adityaomar3) [reply python/avogadro/cjson.py:40]: Ok! I got this, actually we can return the array too, but earlier in the you told (https://github.com/OpenChemistry/avogadrolibs/issues/712#issuecomment-1712893503 point 2) to return the cjson so I have to make it into dictionary element as Cjson interconversion is python dictionaries.

I have a idea that I should return the array and make private to_cjson() fun public which would take array or list as input and convert it into cjson

File: python/avogadro/cjson.py (hunk 5 above)
Comment (adityaomar3) [reply python/avogadro/cjson.py:17]: Sure! I will keep this in mind from now on.

I will check for strings and links as you told 
currently It was for the filepath of file present in the system

File: python/avogadro/cjson.py (hunk 3 above)
Comment (adityaomar3) [reply python/avogadro/cjson.py:26]: What will be the format of param or file for set_atom_coordinates @ghutchis 
only array of coords will be given or array of coords + atomic number

Comment (adityaomar3) [reply python/avogadro/cjson.py:26]: @ghutchis can you please help me with the format for input array of coordinates and elements for set_atoms_coordinates.

**Comment by adityaomar3** [comment]
@ghutchis please review the functions added 
Also, currently the set_atoms_coordinates and set_coordinates sort of work similar, I need a bit help in format of input array for both functions(a sample) to distinguish their functionality

**Comment by ghutchis** [comment]
Okay, I'm going to merge this and tweak it a bit. Thanks.

**Comment by adityaomar3** [comment]
Thankyou sir! This file still needs to be checked on function 
set_atoms_coordinates.
Will create a new patch for that if necessary

**Pull request merged**